src/synthesize.py

```diff
@@ -1,114 +1,136 @@
-#
-# In this algorithm, we're recursively building up larger and larger programs
-# based on our grammar, eargely pruning those that behave identically on all inputs.
-def synthesize(gdfs, target):
-    program_list = []
+def program(gdfs: GdfBindings) -> CandidateGen:
+    'Generates all valid programs (simplest first) over `gdfs`'
+    # generate just the inputs as outputs
+    yield from map(GDF, gdfs.keys())
+    # generate univariate programs
+    yield from univariate(gdfs)
+    # generate bivariate programs
+    yield from bivariate(gdfs)
 
-    for gdf_name in gdfs.keys():
-        program_list.append(GDF(gdf_name))
 
+def univariate(gdfs: dict[str, DataFrame]) -> CandidateGen:
+    'Generates all valid univariate programs over `gdfs`'
     for gdf_name, gdf in gdfs.items():
         for col in gdf.columns:
-            if col != "geometry":
-                program_list.append(Dissolve(GDF(gdf_name), col))
-
-    # Binary operations here.
-
-    print(
-        "Candidates\n",
-        "\n".join([str(program) for program in program_list]),
-    )
-
-    program_list = elim_equivalents(program_list, gdfs)
-
-    # grow will go here if we end up having non-terminals.
-
-    for program in program_list:
-        if gdfs_equal(evaluate_program(program, gdfs), target):
-            print("\nFound it!", program)
-            return program
+            if col != 'geometry':  # geometry cannot be dissolved
+                yield Dissolve(gdf_name, by=col)
 
-    print("No program found!")
 
+def bivariate(gdfs: GdfBindings) -> CandidateGen:
+    ''' Generates all valid bivariate programs over `gdfs`
 
-from geopandas import GeoDataFrame
-
-def lazy_synthesize_all(gdfs: dict[str, GeoDataFrame]):
-    'Generates the simplest programs first, including equivalents.'
-    for gdf_name in gdfs.keys():
-        yield GDF(gdf_name)
+        Due to external implementations, ordering is pseudo-random
+    '''
+    # interleave results to attempt to maximize search space coverage
+    merge_gen = merge(gdfs)
+    sjoin_gen = sjoin(gdfs)
+    # zip stops when one of its sources ends
+    for m_candidate, s_candidate in zip(merge_gen, sjoin_gen):
+        yield m_candidate
+        yield s_candidate
+    # yield whatever remains of whichever still has values left
+    yield from sjoin_gen
+    yield from merge_gen
 
-    for gdf_name, gdf in gdfs.items():
-        for col in gdf.columns:
-            if col != 'geometry':
-                yield Dissolve(GDF(gdf_name), col)
 
-
-def make_candidate_filter(gdfs: dict[str, GeoDataFrame], target: GeoDataFrame):
+def make_candidate_filter(gdfs: GdfBindings, target: DataFrame) -> Callable[[Candidate], bool]:
     'Returns a predicate that checks if a `program` over `gdfs` matches `target`'
-    return lambda program: gdfs_equal(evaluate_program(program, gdfs), target)
+    return lambda program: GeoDataFrame.equals(program.interpret(gdfs), target)
```

Comment:
I think we'll still want to use [`assert_geodataframe_equal`](https://geopandas.org/en/stable/docs/reference/api/geopandas.testing.assert_geodataframe_equal.html#geopandas.testing.assert_geodataframe_equal) from `geopandas` for dataframe equivalence checking. A big reason for this is the `check_crs` parameter, which will verify that the `GeoDataFrame` produced by our program is in the same coordinate reference system as the output. This will be important when we add synthesis support for reprojection, which is an incredibly common spatial analysis task but one participants forget to do often!

Reply:
Gotcha. I wonder if there's a way around using an assert because the creation of try/except blocks is big performance hit (about +25%). Reading the geopandas source [here](https://github.com/geopandas/geopandas/blob/be647ab7e54880278cba374cc77075afd33b25a5/geopandas/testing.py#L239), it looks like we'd need to write our own version of it. What do you think?

Reply:
Ah gotcha! In that case, I would just steal and adapt [these lines](https://github.com/geopandas/geopandas/blob/be647ab7e54880278cba374cc77075afd33b25a5/geopandas/testing.py#L292-L295) checking the CRS equivalence. In essence, we _never_ want to be doing any types of spatial operations on `GeoDataFrames` that aren't in the same spatial reference, but it also seems we can get just this check without the heavy runtime hit of `try` / `except`. I didn't realize they were such a perf hit 😱

Reply:
I've been reviewing those lines and I realized another reason why I switched - it often reports that nothing is equal. I've been reading it and messing with it and I haven't isolated why, just that it performs and behaves completely different from `GeoDataFrame.equals`. In fact, reverting the code to the old equality checker makes the synthesizer instantly fail the univariate synthesis task...

Reply:
I tried to adapt that method by delaying the use of try/except as much as possible. It worked, but it's ugly, and the performance gains are minor, less than 10%. Let me know if you want me to go back and just use one try/except around `assert_geoframe_equals`. Otherwise we should be good to go!

Reply:
I don't think it's worth the perf optimization to bring in all of this code at this stage, so let's stick with `assert_geodataframe_equal` for now, and in the long run of this project I'll know to focus on this area for optimization.